### DISCUSSION TITLE
What is a good position to place bundled files in HTML?

### DISCUSSION THREAD

**Discussion by adavisievolve** [datenstrom/community#790] [Ask a question] [answered]
The Bundle extension will bundle css and js in the **head**, into a js and css file in the head. But some files (js) would be better placed at the end of the page before the closing **body** tag, because they are not needed until later in the page load.

It would be nice if bundle extension had an additional data-attrubute to specify the location the final bundle file(s) should reside. This would allow the possibility of multiple bundle files created, to be included in the **head** or at the end of the page ( perhaps via $this->yellow->page->getExtra("footer") )

`data-location="[head|end]"`

Website with unbundled CSS and JavaScript files:

```
<!DOCTYPE html>
<html>
<head>
<title>Example page</title>
<link rel="stylesheet" type="text/css" media="all" href="/media/extensions/gallery.css" />
<link rel="stylesheet" type="text/css" media="all" href="/media/extensions/twitter.css" />
<link rel="stylesheet" type="text/css" media="all" href="/media/themes/stockholm.css" />
<script type="text/javascript" defer="defer" data-location="end" src="/media/extensions/gallery-photoswipe.min.js"></script>
<script type="text/javascript" defer="defer" data-location="end" src="/media/extensions/gallery.js"></script>
<script type="text/javascript" defer="defer" src="/media/extensions/twitter.js"></script>
</head>
<body>
<h1>Hello world</h1>
</body>
</html>
```

Website with bundled CSS and JavaScript files:

```
<!DOCTYPE html>
<html>
<head>
<title>Example page</title>
<link rel="stylesheet" type="text/css" media="all" href="/media/extensions/bundle-dfd1ef8a4c.min.css" />
<script type="text/javascript" defer="defer" src="/media/extensions/bundle-1.min.js"></script>
</head>
<body>
<h1>Hello world</h1>
<script type="text/javascript" defer="defer" src="/media/extensions/bundle-2.min.js"></script>
</body>
</html>
```

Thoughts?

**Comment by schulle4u** [comment, marked as answer]
The bundle extension defers script loading, all javascripts will be executed as soon as the DOM is fully loaded. Therefore in most cases it shouldn't make a difference if you put some scripts in the body. Some js modules might need this for technical reasons, but this can be solved individually in each extension. 

For more information on this topic see also datenstrom/community#289

**Reply by adavisievolve** [reply]
Oh, I see. Thank you for your quick and kind reply.
I had run a comparison of loading a page with and without the bundler enabled, and it was about the same speed (fewer requests with bundler). I think the css/js are already minified and optimized, so the only difference is fewer requests, which compared to everything else makes a marginal difference in this case. 
Cheers - Andrew
